### ISSUE TITLE
Rollback weak variables analysis and weak variable binding

### ISSUE THREAD

**Issue by jburse** [jburse/jekejeke-devel#220]
Interestingly currently clause heads such as below, contain
two potential weak variables H and T. At least if walk is called with
a ('.')/2 compound, then we can consider H and T weak:
```
walk([H|T]) :- walk(H), walk(T).
```
Nice coincidence, we consider weak variables intially strong. So 
we could furnish a unification which can turn variables from strong 
into weak on the go. The best would be if this information is available 

inside the compound. The result of a used analysis fused with the 
body analysis is needed. So we might first do the body analysis and then 
the fused analysis. We could adapt the compound adornment code

which we deviced for the linear/mixed unification.

**Comment by jburse** [comment]
Interestingly some head structure occurence is no obstacle.
We can also have head structure occurence and still turn variables 
from strong into weak on the go. Simply check the reference count

of the variable. If the variable is fresh, we can flip it otherwise not.
So we dont need static used analysis  information since we have this 
information dynamically available at run time. We only need the body 

analysis statically available. But for this we could introduce a new Java 
class **SkelVarMarkable**, similar like we already introduce 
**DisplayMarkable**. This type of variable would be generated in read 

term new for the boot loader and in copy term new when clauses are 
prepared. Ultimately it could even save us the Optimization helper 
through clever coding. We would then have a new unification routine 

that rely on this information. Checking whether this new unification 
routine can be called is then as easy as inspecting the variable string.

**Comment by jburse** [comment]
But maybe we are solving the wrong problem. This ticked was
targetted towards examples such as:
```
norm(A, R) :- var(A), !, R = pos(A).
norm((A;B), R) :- !, norm(A, C), norm(B, D), R = or(C,D).
norm((A,B), R) :- !, norm(A, C), norm(B, D), R = and(C,D).
Etc...
```
But there is another problem with this code. Occurs check will
make it quadratic when not optimized away, since the `R=or(C,D)`
are called when `C` and `D` are available.

But putting `R=or(C,D)` at the end also destroys recursive last call 
optimization. So the recursive norm/2 calls will not have their frames 
removed. A better coding would be:
```
norm(A, R) :- var(A), !, R = pos(A).
norm((A;B), R) :- !, R = or(C,D), norm(A, C), norm(B, D).
norm((A,B), R) :- !, R = and(C,D), norm(A, C), norm(B, D).
Etc...
```
The code will be linear in terms of occurs check, and recursive 
last call optimization is in place again.

**Comment by jburse** [comment]
The rollback has been implemented.

**Comment by jburse** [comment]
Rollback was done wrongly. We didn't notice that we still need the
weak variables from the display, and that the old display implementation
does not work with the dynamic heuristic to drop the occurs check.

We only noticed that things go wrong, when we did some regression
testing of Quine Scheme. Both explicit unify_with_occurs_check went
wrong and occurs_check=true went wrong.

But it has now been fixed. Question is do we have test cases?
